### PR TITLE
SEM udpates

### PR DESCRIPTION
**Pull request by SachaEpskamp**
SEM updates: Multiple models, covariance matrix as input, measurement
invariance tests.

### CODE REVIEW

File: JASP-Engine/JASP/R/commonerrorcheck.R [hunk 1]
```diff
@@ -456,4 +457,28 @@
     
   }
   return(result)
+}
+
+# Check if data set is variance-covariance matrix
+.checkVarCovMatrix <- function(dataset,nrow=TRUE,symm=TRUE,posdef=TRUE,...){
+
+  # as matrix:
+  dataMatrix <- as.matrix(dataset)
```
Comment (AlexanderLyNL) [inline JASP-Engine/JASP/R/commonerrorcheck.R]: I don't see dataMatrix being used in the rest of this function

File: JASP-Engine/JASP/R/commonerrorcheck.R [hunk 2]
```diff
@@ -456,4 +457,28 @@
...
+    return(list(error=TRUE,reason="Dataset is not a square matrix"))
+  }
+  
+  # Symmetrical?
+  if (symm && !all(round(dataset,10) == t(round(dataset,10)))){
+    return(list(error=TRUE,reason="Matrix is not symmetrical"))
+  }
+  
+  # Positive-definite?
+  if (posdef && any(round(eigen(dataset)$values,10) < 0)){
```
Comment (AlexanderLyNL) [inline JASP-Engine/JASP/R/commonerrorcheck.R]: In general, this is necessary, but not a sufficient condition.

File: JASP-Engine/JASP/R/commonerrorcheck.R [hunk 3]
```diff
@@ -456,4 +457,28 @@
...
+    return(list(error=TRUE,reason="Matrix is not positive-definite"))
```
Comment (AlexanderLyNL) [inline JASP-Engine/JASP/R/commonerrorcheck.R]: space after comma

File: JASP-Engine/JASP/R/commonerrorcheck.R [hunk 4]
```diff
@@ -456,4 +457,28 @@
     
   }
   return(result)
+}
+
+# Check if data set is variance-covariance matrix
+.checkVarCovMatrix <- function(dataset,nrow=TRUE,symm=TRUE,posdef=TRUE,...){
+
```
Comment (AlexanderLyNL) [inline JASP-Engine/JASP/R/commonerrorcheck.R]: spaces after commas

File: JASP-Engine/JASP/R/commonerrorcheck.R [hunk 5]
```diff
@@ -456,4 +457,28 @@
...
+  # as matrix:
+  dataMatrix <- as.matrix(dataset)
+  
+  # number of rows equal to number of columns?
+  if (nrow && nrow(dataset) != ncol(dataset)){
+    return(list(error=TRUE,reason="Dataset is not a square matrix"))
```
Comment (AlexanderLyNL) [inline JASP-Engine/JASP/R/commonerrorcheck.R]: space after comma

File: JASP-Engine/JASP/R/commonerrorcheck.R [hunk 6]
```diff
@@ -456,4 +457,28 @@
...
+  }
+  
+  # Symmetrical?
+  if (symm && !all(round(dataset,10) == t(round(dataset,10)))){
+    return(list(error=TRUE,reason="Matrix is not symmetrical"))
```
Comment (AlexanderLyNL) [inline JASP-Engine/JASP/R/commonerrorcheck.R]: spaces after commas. "Matrix is not symmetric" vs "Matrix is not symmetrical"

File: JASP-Engine/JASP/R/commonmessages.R [hunk 7]
```diff
@@ -27,6 +27,9 @@
   messages$error$observations <- "Number of observations {{observations.amount}} in {{variables}}"
   messages$error$levene <- "Cannot compute statistic reliably: number of observations {{observations.amount}} in {{variables}}"
   
+  messages$error$varCovMatrix <- "Dataset is not a proper variance-covariance matrix. Please load only a positive definite symmetrical matrix as your dataset."
```
Comment (AlexanderLyNL) [inline JASP-Engine/JASP/R/commonmessages.R]: Last sentence sounds weird to me and might even be redundant.

File: JASP-Desktop/ribbons/ribbonsem.ui [hunk 8]
```diff
@@ -161,6 +161,7 @@
  </customwidgets>
  <resources>
   <include location="../resources/icons.qrc"/>
+  <include location="../resources/icons.qrc"/>
```
Comment (akashrajkn) [inline JASP-Desktop/ribbons/ribbonsem.ui]: Please remove the duplicated statement.

File: JASP-Engine/JASP/R/exploratoryfactoranalysis.R [hunk 9]
```diff
@@ -215,14 +215,16 @@ ExploratoryFactorAnalysis <- function(dataset = NULL, options, perform = "run",
...
   if (perform == "run" && status$ready) {
     state <- list(options=options,analysisResults=analysisResults,nFactor=nFactor,results=results,complete=TRUE)
-    
-    return(list(results=results, status="complete", state=state))	    
-    
-    
+    retList <- list(results=results, status="complete", state=state)
+
   } else {
     state <- list(options=options,nFactor=nFactor,analysisResults=analysisResults,results=results,complete=FALSE)
```
Comment (AlexanderLyNL) [inline JASP-Engine/JASP/R/exploratoryfactoranalysis.R]: comma space

File: JASP-Engine/JASP/R/exploratoryfactoranalysis.R [hunk 10]
```diff
@@ -215,14 +215,16 @@ ExploratoryFactorAnalysis <- function(dataset = NULL, options, perform = "run",
...
-    return(list(results=results, status="inited",state=state))
+    
+    retList <- list(results=results, status="inited",state=state)
```
Comment (AlexanderLyNL) [inline JASP-Engine/JASP/R/exploratoryfactoranalysis.R]: comma space

File: JASP-Engine/JASP/R/semsimple.R [hunk 11]
```diff
@@ -21,52 +21,52 @@
...
 	matList <- mapply(cov = covList, cor = corList, type = names(covList), FUN=function(cov,cor, type){
-		cov[upper.tri(cov,diag=FALSE)] <- cor[upper.tri(cor,diag=FALSE)] 
+		cov[upper.tri(cov,diag=FALSE)] <- cor[upper.tri(cor,diag=FALSE)]
 		cbind(..sortingDummy = seq_len(NROW(cov)), ..varName = .unv(rownames(cov)), ..type = type, as.data.frame(cov),stringsAsFactors = FALSE)
 	}, SIMPLIFY = FALSE)
-	
+
 	matDF <- do.call(rbind, matList)
 	matDF <- matDF[matDF$..type %in% include,]
 	matDF$..type <- as.character(	matDF$..type)
```
Comment (AlexanderLyNL) [inline JASP-Engine/JASP/R/semsimple.R]: ( too many spaces

File: JASP-Engine/JASP/R/commonerrorcheck.R (hunk 2 above)
Comment (quentingronau) [inline JASP-Engine/JASP/R/commonerrorcheck.R]: Hi Sacha,

This is a very minor thing and probably not relevant in practice, but it looks like you are checking whether the matrix is positive-semidefinite and not positive-definite? So should this be "<= 0" instead?

Cheers,
Quentin

Comment (SachaEpskamp) [inline JASP-Engine/JASP/R/commonerrorcheck.R]: Yes, but positive-semidefinite is fine enough right?

File: JASP-Desktop/ribbons/ribbonsem.ui (hunk 8 above)
Comment (SachaEpskamp) [inline JASP-Desktop/ribbons/ribbonsem.ui]: Done

File: JASP-Engine/JASP/R/commonerrorcheck.R (hunk 4 above)
Comment (SachaEpskamp) [inline JASP-Engine/JASP/R/commonerrorcheck.R]: Done

File: JASP-Engine/JASP/R/commonerrorcheck.R (hunk 1 above)
Comment (SachaEpskamp) [inline JASP-Engine/JASP/R/commonerrorcheck.R]: Done

File: JASP-Engine/JASP/R/commonerrorcheck.R (hunk 6 above)
Comment (SachaEpskamp) [inline JASP-Engine/JASP/R/commonerrorcheck.R]: Done

File: JASP-Engine/JASP/R/commonerrorcheck.R (hunk 3 above)
Comment (SachaEpskamp) [inline JASP-Engine/JASP/R/commonerrorcheck.R]: Done

File: JASP-Engine/JASP/R/semsimple.R [hunk 12]
```diff
@@ -21,52 +21,52 @@
...
-	
+
 	n <- ncol(covList[[1]])
-	
-	
+
+
 	matList <- mapply(cov = covList, cor = corList, type = names(covList), FUN=function(cov,cor, type){
-		cov[upper.tri(cov,diag=FALSE)] <- cor[upper.tri(cor,diag=FALSE)] 
+		cov[upper.tri(cov,diag=FALSE)] <- cor[upper.tri(cor,diag=FALSE)]
```
Comment (SachaEpskamp) [inline JASP-Engine/JASP/R/semsimple.R]: Done

File: JASP-Engine/JASP/R/semsimple.R [hunk 13]
```diff
@@ -207,35 +207,77 @@
...
 
+
 	variables <- .extractVariables(options$model)
 	model <- .translateModel(options$model, variables)
 
 	if (is.null(dataset)) {
-	
-		if (perform == "run") {
+
+		if (perform == "run" ||options$Data != "varcov") {
```
Comment (SachaEpskamp) [inline JASP-Engine/JASP/R/semsimple.R]: Done

File: JASP-Engine/JASP/R/semsimple.R [hunk 14]
```diff
@@ -207,35 +207,77 @@
...
+	}
+	# Check mean structure:
+	if (options$Data == "varcov"){
+	  if (isTRUE(options$includeMeanStructure)){
+	    errorCheck <- list(error=TRUE,message = "Mean structure can not be included when data is variance-covariance matrix")
+	  }
+	  options$includeMeanStructure <- FALSE
+	  
+	  if (options$SampleSize == 0){
+	    errorCheck <- list(error=TRUE,message = "Please set the sample size!")
```
Comment (SachaEpskamp) [inline JASP-Engine/JASP/R/semsimple.R]: Done

File: JASP-Engine/JASP/R/semsimple.R [hunk 15]
```diff
@@ -207,35 +207,77 @@
...
+	  }
+	  
+	  # Check for multiple groups:
+	  if (options$groupingVariable != ""){
+	    errorCheck <- list(error=TRUE,message = "Multiple group analysis not (yet) supported when data is variance-covariance matrix")
+	  }
+	} else {
+	  if (identical(errorCheck, FALSE)){
+	    if (!nrow(dataset) > ncol(dataset)){
+	      errorCheck <- list(error=TRUE,message = "Not more cases than number of variables. Is your data a variance-covariance matrix?")
```
Comment (SachaEpskamp) [inline JASP-Engine/JASP/R/semsimple.R]: Done

File: JASP-Engine/JASP/R/semsimple.R [hunk 16]
```diff
@@ -244,31 +286,115 @@ SEMSimple <- function(dataset=NULL, options, perform="run", callback=function(..
...
+	    semResults <- try(lavaan:::lavaan(
+	      model = model,
+	      data = dataset,
+	      auto.delta = options$addScalingParameters,
+	      auto.th = options$addThresholds,
+	      orthogonal = options$assumeFactorsUncorrelated,
+	      auto.cov.y = options$correlateDependentVariables,
+	      auto.cov.lv.x = options$correlateExogenousLatents,
+	      mimic = ifelse(options$emulation=="none","lavaan",options$emulation),
+	      se = options$errorCalculation,
```
Comment (SachaEpskamp) [inline JASP-Engine/JASP/R/semsimple.R]: Done

File: JASP-Engine/JASP/R/semsimple.R [hunk 17]
```diff
@@ -244,31 +286,115 @@ SEMSimple <- function(dataset=NULL, options, perform="run", callback=function(..
...
+	    semResults <- try(lavaan:::lavaan(
+	      model = model,
+	      sample.cov = as.matrix(dataset),
+	      sample.nobs = as.numeric(options$SampleSize),
+	      auto.delta = options$addScalingParameters,
+	      auto.th = options$addThresholds,
+	      orthogonal = options$assumeFactorsUncorrelated,
+	      auto.cov.y = options$correlateDependentVariables,
+	      auto.cov.lv.x = options$correlateExogenousLatents,
+	      mimic = ifelse(options$emulation=="none","lavaan",options$emulation),
```
Comment (SachaEpskamp) [inline JASP-Engine/JASP/R/semsimple.R]: Done

File: JASP-Engine/JASP/R/commonerrorcheck.R (hunk 2 above)
Comment (quentingronau) [inline JASP-Engine/JASP/R/commonerrorcheck.R]: Yes, would it then maybe be more precise to change the reason to
reason="Matrix is not positive-semidefinite"?

# Positive-definite?
  if (posdef && any(round(eigen(dataset)$values,10) < 0)){
    return(list(error=TRUE,reason="Matrix is not positive-definite"))
  }

**Pull request merged**